cms/djangoapps/contentstore/rest_api/v0/serializers/tabs.py

```diff
@@ -0,0 +1,109 @@
+    )
+    name = serializers.CharField(
+        read_only=True,
+        help_text=_("Name of the tab displayed to users. Overrides title."),
+    )
+    tab_id = serializers.CharField(
+        read_only=True,
+        help_text=_("Name of the tab displayed to users. Overrides title."),
+    )
+    settings = serializers.DictField(
+        read_only=True,
+        help_text=_("Additional settings specific to the tab"),
+    )
+
+    def to_representation(self, instance: CourseTab):
+        tab_data = {
+            "type": instance.type,
+            "title": instance.title,
+            "is_hideable": instance.is_hideable,
+            "is_hidden": instance.is_hidden,
+            "is_movable": instance.is_movable,
+            "course_staff_only": instance.course_staff_only,
+            "name": instance.name,
+            "tab_id": instance.tab_id,
+        }
+        tab_settings = {
+            key: value for key, value in instance.tab_dict.items() if key not in tab_data and key != "link_func"
+        }
+        tab_data["settings"] = tab_settings
+        return tab_data
+
+
+class TabIDLocatorSerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """Serializer for tab locations, used to identify a tab in a course."""
+
+    tab_id = serializers.CharField(required=False, help_text=_("ID of tab to update"))
+    tab_locator = UsageKeyField(required=False, help_text=_("Location (Usage Key) of tab to update"))
+
+    def validate(self, attrs):
+        if "tab_id" not in attrs and "tab_locator" not in attrs:
+            raise serializers.ValidationError("At least one of tab_id and tab_locator is required")
+        return super().validate(attrs)
+
+
+class CourseTabUpdateSerializer(serializers.Serializer):  # pylint: disable=abstract-method
+    """
+    Serializer to update course tabs.
+    """
+
+    tab_id_locator = TabIDLocatorSerializer(
+        required=False,
+        help_text=_("A tab location expressed via either a `tab_id` or `tab_location`"),
+    )
+    is_hidden = serializers.BooleanField(
+        required=False,
+        help_text=_("True to hide the tab, and False to show it."),
+    )
+    tabs = serializers.ListField(
+        required=False,
+        child=TabIDLocatorSerializer(help_text=_("A tab location expressed via either a `tab_id` or `tab_location`")),
+        help_text=_("A list of tab locators in the order in which they should appear in the UI."),
+        min_length=2,
+    )
+
+    def validate(self, attrs):
+        """Validate the the correct combination of fields has been provided via the API."""
+        tab_id_locator = "tab_id_locator" in attrs
+        is_hidden = "is_hidden" in attrs
+        tabs = "tabs" in attrs
+        if (
+            # Either supply a `tabs` field or the `tab_id_locator` and `is_hidden` fields
+            (tabs and (is_hidden or tab_id_locator))
+            or (not tabs and not (is_hidden and tab_id_locator))
+        ):
```

Comment:
That looks much cleaner although I've switched around the logic a bit.

Reply:
This does make me feel like perhaps there should be two update endpoints. One to reorder, and the other to update, since these feel like very different operations. Otherwise we should have a unified update and format that does both.

Reply:
I agree that separate endpoints would be the best choice but did not suggest it because the scope here is to just expose what already exists... my experience with such changes is that they explode very quickly.
This is something I can live with.

Reply:
I don't think your code sample has the same logic. In your code sample it will be considered an update if either of `is_hidden` or `tab_id_locator` is provided in the absence of `tabs` a validation error won't be raised. I'll keep the original code.

Reply:
You're right, it should have been `set(attrs) >= {"tab_id_locator", "is_hidden"}` or something similar.